File: Python/DFS/BinaryTreePathSum.py
# @author
# Aakash Verma

# Output:
# true

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class BinaryTreePathSum:
    
    def __init__(self): 
        self.root = None

    def hasPath(self, root, sum):
        if(root is None): return False

        if(root.data == sum and root.left is None and root.right is None):
            return True

        return self.hasPath(root.left, sum - root.data) or self.hasPath(root.right, sum - root.data)
        
if __name__ == '__main__':
    tree = BinaryTreePathSum()
    tree.root = Node(12)
    tree.root.left = Node(7)
    tree.root.right = Node(1)
    tree.root.left.left = Node(9)
    tree.root.right.left = Node(10)
    tree.root.right.right = Node(5)
    tree.root.left.left.left = Node(3)
    ans = tree.hasPath(tree.root, 23)
    print(ans)

File: Java/DFS/BinaryTreePathSum.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   true
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class BinaryTreePathSum {

    static Node root;
    BinaryTreePathSum() {
        root = null;
    }

    public static boolean hasPath(Node root, int sum) {
        if(root == null) return false;

        if(root.data == sum && root.left == null && root.right == null)
            return true;

        return hasPath(root.left, sum - root.data) || hasPath(root.right, sum - root.data);
    }

    public static void main (String[] args) {
        root = new Node(12);
        root.left = new Node(7);
        root.right = new Node(1);
        root.left.left = new Node(9);
        root.right.left = new Node(10);
        root.right.right = new Node(5);
        root.left.left.left = new Node(3);
        boolean ans = hasPath(root, 23);
        System.out.println(ans);
    }
}